### PR TITLE
refactor(api): Represent option dicts as a structs in C and reduce conversion overhead from lua

### PR DESCRIPTION
**Pull request by bfredl**
Problem: using a `Dictionary opts` is all the more common in API functions for optional args.
However accessing these from the C side is much clumsier than just accessing a positional arg.
This is most apparent in `parse_float_config` where we need to keep a lot of extra state for which
arguments we've seen so far, instead of directly checking the args in an order that makes sense.

Also there's a notable _performance_ hit when using such API functions from lua, from (1) copying lua key strings into separate allocated API strings and (2) doing `strequal()` in a loop which is `O(N^2)` if all keys are present. 

Solution: instead generate a struct to represent an option-dict with known (but optional) keys, and use a simple hash function to do a one pass `O(N)` conversion of present keys directly from the lua state. As an example, the _overall_ cost of a `vim.api.nvim_buf_set_extmark` call from lua is cut in half (as measured in the inner loop of tree-sitter redrawing, where it is a quite significant part of the redraw time).

As a possible follow up, the present keys could be present in _metadata_. but as that affects interface stability and versioning, it is better to discuss that in a separate PR/issue. this PR only focuses on performance and _c implementation_ of existing interfaces.

### CODE REVIEW

File: src/nvim/api/buffer.c [hunk 1]
```diff
@@ -1481,7 +1481,7 @@ Array nvim_buf_get_extmarks(Buffer buffer, Integer ns_id,
 /// @return Id of the created/updated extmark
 Integer nvim_buf_set_extmark(Buffer buffer, Integer ns_id,
                              Integer line, Integer col,
-                             Dictionary opts, Error *err)
+                             KeyDictionary(set_extmark) *opts, Error *err)
```
Comment (justinmk) [inline src/nvim/api/buffer.c:1484]: can we call it just `Dict` with the plan of removing `Dictionary` entirely?

Comment (bfredl) [reply src/nvim/api/buffer.c:1484]: Yes.

File: src/nvim/api/keysets.lua [hunk 2]
```diff
@@ -0,0 +1,19 @@
+return {
+  set_extmark = {
```
Comment (justinmk) [inline src/nvim/api/keysets.lua:2]: should it be `buf_set_extmark`? Or even the full api name `nvim_buf_set_extmark` , to avoid cognitive indirection.

Comment (bfredl) [reply src/nvim/api/keysets.lua:2]: Not sure, but they are not public and not necessarily tied to one specific function, for instance we have two entry points for float_config.

Comment (justinmk) [reply src/nvim/api/keysets.lua:2]: ah right, these are more like types, not signatures

Comment (bfredl) [reply src/nvim/api/keysets.lua:2]: Currently all key sets are for Object structs, but they could also latter be used for other stuff, like method lookup tables (and replace gperf, hint hint).

Comment (justinmk) [reply src/nvim/api/keysets.lua:2]: > (and replace gperf, hint hint

read my mind

File: src/nvim/lua/converter.c [hunk 3]
```diff
@@ -1299,3 +1299,25 @@ void nlua_init_types(lua_State *const lstate)
...
+void nlua_pop_keydict(lua_State *L, void *retval, field_hash hashy, Error *err)
+{
+  lua_pushnil(L);  // [dict, nil]
+  while (lua_next(L, -2)) {
+    // [dict, key, value]
+    size_t len;
+    const char *s = lua_tolstring(L, -2, &len);
+    Object *field = hashy(retval, s, len);
+    if (!field) {
+      api_set_error(err, kErrorTypeValidation, "unexpected key: %.*s", (int)len, s);
```
Comment (justinmk) [inline src/nvim/lua/converter.c:1313]: "invalid key" is used in most of the API I think
```suggestion
      api_set_error(err, kErrorTypeValidation, "invalid key: %.*s", (int)len, s);
```

File: src/nvim/generators/hashy.lua [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+  local len_pos_buckets = {}
+  local worst_buck_size = 0
+
+  for len = 1,maxlen do
+    local strs = len_buckets[len]
+    if strs then
+      --print("len = "..len..", strs="..#strs)
+      -- the best position so far generates `best_bucket`
+      -- with `minsize` wort case collitions
```
Comment (justinmk) [inline src/nvim/generators/hashy.lua:32]:
```suggestion
      -- with `minsize` wort case collisions
```

Comment (jamessan) [reply src/nvim/generators/hashy.lua:32]: wort -> worst too :)

File: src/nvim/lua/converter.c (hunk 3 above)
Comment (bfredl) [reply src/nvim/lua/converter.c:1313]: not as far as the code is telling me (zero hits on "invalid key")

Comment (bfredl) [reply src/nvim/lua/converter.c]: ok it is inconsistent between the interfaces, changing this one for now

**Pull request merged**